cineast-api/src/main/java/org/vitrivr/cineast/api/util/QueryUtil.java

```diff
@@ -34,39 +42,135 @@
+    var stageConfig = config.clone();
+
+    var stagedQueryResults = new ArrayList<HashMap<String, List<StringDoublePair>>>();
+
+    for (QueryStage stage : stages) {
+      var stageResults = findSegmentsSimilar(continuousRetrievalLogic, stage.terms, stageConfig);
+      stagedQueryResults.add(stageResults);
+
+      var relevantSegments = new HashSet<String>();
+      for (var result : stageResults.values()) {
+        relevantSegments.addAll(result.stream().map(pair -> pair.key).collect(Collectors.toList()));
+      }
+
+      // Return empty results if there are no more results in stage
+      if (relevantSegments.isEmpty()) {
+        return stageResults;
+      }
+
+      stageConfig.setRelevantSegmentIds(relevantSegments);
+    }
+
+    return mergeStagedQueryResults(stagedQueryResults);
+  }
+
+  /**
+   * Executes a temporal similarity query.
+   *
+   * @param continuousRetrievalLogic The continuous retrieval logic to execute the query.
+   * @param query                    The temporal query to execute.
+   * @param config                   The config to use for the execution of the query
+   * @return The query results as a list of temporal objects.
+   */
+  public static List<TemporalObject> findSegmentsSimilarTemporal(ContinuousRetrievalLogic continuousRetrievalLogic, TemporalQuery query, QueryConfig config) {
+    var stagedResults = query.getQueries().stream()
+        .map(stagedQuery -> findSegmentsSimilarStaged(continuousRetrievalLogic, stagedQuery.getStages(), config))
+        .collect(Collectors.toList());
+
+    // TODO: New MediaSegmentReader for every request like FindSegmentByIdPostHandler or one persistent on per endpoint like AbstractQueryMessageHandler?
+    final var segmentReader = new MediaSegmentReader(Config.sharedConfig().getDatabase().getSelectorSupplier().get());
+
+    var segmentIds = stagedResults.stream().flatMap(
+        resultsMap -> resultsMap.values().stream().flatMap(
+            pairs -> pairs.stream().map(pair -> pair.key)
+        )
+    ).distinct().collect(Collectors.toList());
+
+    var segmentDescriptors = segmentReader.lookUpSegments(segmentIds);
+    var stagedQueryResults = stagedResults.stream().map(
+        resultsMap -> resultsMap.values().stream().flatMap(Collection::stream).collect(Collectors.toList())
+    ).collect(Collectors.toList());
+
+    return TemporalScoring.score(segmentDescriptors, stagedQueryResults, query.getTimeDistances(), query.getMaxLength());
+  }
+
+  /**
+   * Merges staged query results into a single result map.
+   *
+   * @param stagedQueryResults List of staged query results by stage. The results are expected to be in order of the stages.
+   */
+  public static HashMap<String, List<StringDoublePair>> mergeStagedQueryResults(ArrayList<HashMap<String, List<StringDoublePair>>> stagedQueryResults) {
```

Comment:
I think the key of the map is the category, right? In this case, this should probably be documented in the Javadoc